### PR TITLE
Add "rush version" to manage version bumps based on version policy.

### PR DESCRIPTION
**Pull request by qz2017**
Adding a "Rush version" verb to manage version bumps based on version policy. 

A version policy file would look like:
[
  {
    "policyName": "myPublic",
    "definitionName": "lockStepVersion",
    "version": "1.20.0-dev.1",
    "nextBump": "prerelease"
  },
  {
    "policyName": "myInternal",
    "definitionName": "individualVersion",
    "lockedMajor": 1
  }
]

Projects in Rush.json with versionPolicyName "myPublic" are my public projects for public sharing.
Projects in Rush.json with versionPolicyName "myInternal" are my internal projects for internal sharing.

Sample commands:
1. rush version --ensure-version-policy
This command will enforce all projects in Rush.json with versionPolicyName "myPublic" to have version 1.20.0-dev.1, and all projects with "myInternal" to have major version 1. 

2. rush version --bump
This command bumps all "myPublic" projects to version 1.20.0-dev.2 and all "myInternal" projects to versions based on change files and major version remains 1.

3. rush version --bump --version-policy "myPublic" --override-bump patch
This command bumps all "myPublic" projects to version 1.20.0.

4. rush publish --include-all 
This command publishes all myPublic projects and myInternal projects.

5. rush publish --include-all --version-policy myPublic
This command publishes all myPublic projects.

### CODE REVIEW

File: apps/rush/src/utilities/ChangeFiles.ts [hunk 1]
```diff
@@ -76,21 +79,51 @@ export default class ChangeFiles {
   /**
    * Delete all change files
    */
-  public deleteAll(shouldDelete: boolean): void {
-    if (this._files.length) {
+  public deleteAll(shouldDelete: boolean, packagesMap?: Map<string, IPackageJson>): number {
```
Comment (qz2017) [inline apps/rush/src/utilities/ChangeFiles.ts]: >packagesMap [](start = 42, length = 11)

should skip the ones that are not in the published version policy as well.

File: apps/rush/src/actions/VersionAction.ts [hunk 2]
```diff
@@ -67,6 +65,18 @@ export default class VersionAction extends BaseRushAction {
       parameterShortName: '-p',
       description: 'The name of the version policy'
     });
+    this._overwriteBump = this.defineStringParameter({
+      parameterLongName: '--override-bump',
+      description: 'Overrides the bump type in the version-policy.json for the specifiedd version policy.' +
```
Comment (pgonzal) [inline apps/rush/src/actions/VersionAction.ts]: >specifiedd [](start = 79, length = 10)

spelling

File: apps/rush/src/actions/VersionAction.ts [hunk 3]
```diff
@@ -77,82 +87,60 @@ export default class VersionAction extends BaseRushAction {
...
-        git.pull();
-        git.merge(tempBranch);
-        git.push(this._targetBranch.value);
-        git.deleteBranch(tempBranch);
+        this._gitProcess(tempBranch);
       }
+    } else if (this._bumpVersion.value) {
+      const tempBranch: string = 'version/bump-' + new Date().getTime();
+      this._versionManager.bump(this._versionPolicy.value,
+        BumpType[this._overwriteBump.value],
```
Comment (pgonzal) [inline apps/rush/src/actions/VersionAction.ts]: >BumpType[this._overwriteBump.value] [](start = 8, length = 35)

what if the value doesn't match anything in the enum?

File: apps/rush/src/actions/VersionAction.ts [hunk 4]
```diff
@@ -77,82 +87,60 @@ export default class VersionAction extends BaseRushAction {
...
-        const change: IChangeInfo = this._createChangeInfo(newPackageJson, rushProject);
-        ChangelogGenerator.updateIndividualChangelog(change,
-          rushProject.projectFolder,
-          true);
-      }
-      // TODO: if only package dependency changes, add change file.
-    });
+  private _getUserEmail(): string {
+    return Utilities.executeCommandAndCaptureOutput('git',
+        ['config', 'user.email'], '.').trim();
```
Comment (pgonzal) [inline apps/rush/src/actions/VersionAction.ts]: @nickpape-msft for Evergreen did you find a library that abstracts our git operations?  We should really try to get away from the practice of executing whatever Git is found in the current path, without any version checking or formal protocol.

File: apps/rush/src/utilities/PublishUtilities.ts [hunk 5]
```diff
@@ -139,7 +150,7 @@ export default class PublishUtilities {
   }
 
   public static isRangeDependency(version: string): boolean {
-    const LOOSE_PKG_REGEX: RegExp = />=?(?:\d+\.){2}\d+\s+<(?:\d+\.){2}\d+/;
+    const LOOSE_PKG_REGEX: RegExp = />=?(?:\d+\.){2}\d+(\-[0-9A-Za-z-.]*)?\s+<(?:\d+\.){2}\d+/;
```
Comment (pgonzal) [inline apps/rush/src/utilities/PublishUtilities.ts]: > />=?(?:\d+\.){2}\d+(\-[0-9A-Za-z-.]*)?\s+<(?:\d+\.){2}\d+/; [](start = 35, length = 60)

:-o

Can you use [semver.validRange()](https://www.npmjs.com/package/semver) for this?

This is turning into a pretty complicated RegExp...  Not sure the 5 unit test cases are really checking all the edge cases.

File: common/reviews/api/rush-lib.api.ts [hunk 6]
```diff
@@ -59,18 +59,24 @@ enum BumpType {
   // (undocumented)
   'minor',
   // (undocumented)
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: >// (undocumented) [](start = 2, length = 17)

Add the missing documentation for all these

File: common/reviews/api/rush-lib.api.ts [hunk 7]
```diff
@@ -59,18 +59,24 @@ enum BumpType {
...
 
 // @public
 class ChangeFile {
   public constructor(private _changeFileData: IChangeFile,
       private _rushConfiguration: RushConfiguration);
+  // (undocumented)
+  public addChange(data: IChangeInfo): void;
   public generatePath(): string;
+  // (undocumented)
+  public getChanges(packageName: string): IChangeInfo[];
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: Missing documentation

File: common/reviews/api/rush-lib.api.ts [hunk 8]
```diff
@@ -198,9 +204,15 @@ interface ILockStepVersionJson extends IVersionPolicyJson {
...
   constructor(versionPolicyJson: IIndividualVersionJson);
   // (undocumented)
+  public bump(bumpType?: BumpType, identifier?: string): void;
+  // (undocumented)
   public ensure(project: IPackageJson): IPackageJson | undefined;
   // (undocumented)
+  public readonly json: IIndividualVersionJson;
+  // (undocumented)
   public readonly lockedMajor: number | undefined;
+  // (undocumented)
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: Missing documentation

File: common/reviews/api/rush-lib.api.ts [hunk 9]
```diff
@@ -265,10 +277,16 @@ class JsonSchemaValidator {
...
   constructor(versionPolicyJson: ILockStepVersionJson);
   // (undocumented)
+  public bump(bumpType?: BumpType, identifier?: string): void;
+  // (undocumented)
   public ensure(project: IPackageJson): IPackageJson | undefined;
   // (undocumented)
+  public readonly json: ILockStepVersionJson;
+  // (undocumented)
   public readonly nextBump: BumpType;
   // (undocumented)
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: Missing documentation

File: common/reviews/api/rush-lib.api.ts [hunk 10]
```diff
@@ -475,6 +495,7 @@ class VersionControl {
   public static getChangedFiles(prefix?: string, targetBranch?: string): string[];
   // (undocumented)
   public static getChangedFolders(targetBranch?: string): string[];
+  public static getUncommittedChanges(): string[];
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: >string[]; [](start = 41, length = 9)

These should all be ReadonlyArray<string>, not string[]

File: common/reviews/api/rush-lib.api.ts [hunk 11]
```diff
@@ -496,19 +517,30 @@ class VersionMismatchFinder {
...
   public readonly definitionName: VersionPolicyDefinitionName;
   // (undocumented)
   public abstract ensure(project: IPackageJson): IPackageJson | undefined;
   // (undocumented)
+  public readonly json: IVersionPolicyJson;
+  // (undocumented)
   public static load(versionPolicyJson: IVersionPolicyJson): VersionPolicy;
   // (undocumented)
   public readonly policyName: string;
+  // (undocumented)
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: Missing documentation

File: common/reviews/api/rush-lib.api.ts [hunk 12]
```diff
@@ -496,19 +517,30 @@ class VersionMismatchFinder {
...
 }
 
 // @alpha (undocumented)
 class VersionPolicyConfiguration {
   public constructor(private _jsonFileName: string);
   // @alpha
+  public bump(versionPolicyName?: string,
+      bumpType?: BumpType,
+      identifier?: string,
+      shouldCommit?: boolean): void;
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: This is too many optional parameters... the usage will be very confusing.  It would be better to use an IVersionPolicyBumpOptions interface

File: common/reviews/api/rush-lib.api.ts [hunk 13]
```diff
@@ -496,19 +517,30 @@ class VersionMismatchFinder {
 class VersionPolicy {
   constructor(versionPolicyJson: IVersionPolicyJson);
   // (undocumented)
+  public abstract bump(bumpType?: BumpType, identifier?: string): void;
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: >bumpType?: BumpType, identifier?: string [](start = 23, length = 40)

Why are these optional?  It would be better to make them required parameters that can be "string|undefined", to ensure that callers don't forget to pass along one of the parameters.

File: apps/rush-lib/src/data/VersionPolicyConfiguration.ts [hunk 14]
```diff
@@ -64,6 +64,39 @@ export class VersionPolicyConfiguration {
     return this._versionPolicies;
   }
 
+  /**
+   * Bumps up versions for the specified version policy or all version policies
+   * @alpha
```
Comment (pgonzal) [inline apps/rush-lib/src/data/VersionPolicyConfiguration.ts]: > * @alpha [](start = 2, length = 9)

If the class is @alpha, then you don't need to mark the members as @alpha

File: apps/rush/src/actions/VersionAction.ts (hunk 4 above)
Comment (qz2017) [inline apps/rush/src/actions/VersionAction.ts]: agree. This is not the only place. We will do another checkin to sort those out.

---
In reply to: [134068909](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134068909) [](ancestors = 134068909)

File: apps/rush/src/utilities/PublishUtilities.ts (hunk 5 above)
Comment (qz2017) [inline apps/rush/src/utilities/PublishUtilities.ts]: The logic here is only for comparator range, not for all types of ranges. We can improve our range support in the next improvement. 

---
In reply to: [134069328](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134069328) [](ancestors = 134069328)

File: apps/rush/src/actions/VersionAction.ts (hunk 3 above)
Comment (qz2017) [inline apps/rush/src/actions/VersionAction.ts]: It will be undefined. Adding a validate method to avoid.

---
In reply to: [134068665](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134068665) [](ancestors = 134068665)

File: common/reviews/api/rush-lib.api.ts (hunk 13 above)
Comment (qz2017) [inline common/reviews/api/rush-lib.api.ts]: These are optional because these are only used to override what policy has defined. 

---
In reply to: [134069787](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134069787) [](ancestors = 134069787)

File: common/reviews/api/rush-lib.api.ts (hunk 12 above)
Comment (qz2017) [inline common/reviews/api/rush-lib.api.ts]: Four parameters are not ideal. But I will leave it for alpha apis. 

---
In reply to: [134069635](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134069635) [](ancestors = 134069635)

File: common/reviews/api/rush-lib.api.ts (hunk 10 above)
Comment (qz2017) [inline common/reviews/api/rush-lib.api.ts]: Fixed the new one. We can fix other APIs when doing major upgrade of Rush.

---
In reply to: [134069569](https://github.com/Microsoft/web-build-tools/pull/277#discussion_r134069569) [](ancestors = 134069569)
